### PR TITLE
Extend run_example.sh to setup App Service, and to build-and-test simple_app_under_app_service.

### PR DESCRIPTION
**Pull request by gapisback**
This commit adds support setup and then build-and-test simple_app_under_app_service .

This requires a multi-step workflow:

- Setup of application_service
- Setup of simple_app_under_app_service
- Starting the Certifier & Application Services
- Running the app in conjunction with the Application Service.

New usage interfaces and the workflow is as follows:
```shell
$ ./run_example.sh application_service setup
$ ./run_example.sh simple_app_under_app_service setup
$ ./run_example.sh --no-cleanup application_service start
$ (Manually kill the Certifier Service started for application_service)
$ ./run_example.sh simple_app_under_app_service run_test
```

Changes include:

- Major modularization and refactoring of run_example.sh
- Streamlined / improved help / usage messages
- Added debug code to few src files to trace error conditions
- Updated various docs / md-files to reflect new workflow
- build.yml:
   - Add new tests to invoke above sequence of commands
   - Refactored test cases into granular named-'run' chunks

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -17,16 +17,18 @@ jobs:
       run: |
         sudo apt update -y
         sudo apt install -y libgtest-dev libgflags-dev openssl libssl-dev protobuf-compiler protoc-gen-go golang-go cmake
-    - name: make
+
+    - name: test-core-certifier-programs
```
Comment (gapisback) [inline .github/workflows/build.yml:21]: Refactored existing tests in build.yml into named chunks of 'run'-workflows. This makes testing and results more granular.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -94,19 +112,95 @@ jobs:
...
+        set -x
+        ps -ef | grep -E 'simpleserver|example_app.exe|run_example.sh|app_service.exe'
+        set +x
+
         ./run_example.sh simple_app
         ./cleanup.sh
 
         popd
 
+    - name: test-build-and-setup-App-Service-and-Simple-App
```
Comment (gapisback) [inline .github/workflows/build.yml:145]: Main new test to exercise the code under this integration.

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -94,19 +112,95 @@ jobs:
...
+        #! Allow some time for App Service to get Certified ...
+        sleep 10
+
+        #! Once Application Service has been certified, we no longer need
+        #! this Certifier Service. Kill it, so app itself can start its
+        #! own Certifier Service.
+        ps -ef | grep simpleserver
+
+        set -x
+        kill -9 $(pgrep simpleserver)
```
Comment (gapisback) [inline .github/workflows/build.yml:179]: FYI: @jlmucb : I tried the alternative of using `--port` options ... but somehow the combinations of arguments needed to get Certifier-specific port numbers did not, quite, work out right.

For now, I'd like to stay with this `kill` approach. And will rework once this change stabilizes in `/main`.

File: application_service/app_service.cc [hunk 4]
```diff
@@ -358,7 +358,8 @@ void app_service_loop(spawned_children* kid, int read_fd, int write_fd) {
   bool continue_loop = true;
 
 #ifdef DEBUG
-  printf("\napplication service loop: %d %d\n", read_fd, write_fd);
+  printf("[%d] Application Service loop: read_fd=%d write_fd=%d\n",
+         __LINE__, read_fd, write_fd);
```
Comment (gapisback) [inline application_service/app_service.cc:362]: This, and other `.cc` files, have similar changes to elaborate diagnostic messages for improved troubleshooting.

File: application_service/app_service.cc [hunk 5]
```diff
@@ -487,26 +495,30 @@ bool process_run_request(run_request& req) {
...
+
+  int fd2[2];
   if (pipe2(fd2, O_DIRECT) < 0) {
     printf("Pipe 2 failed\n");
     return false;
   }
 
-#ifdef DEBUG
-  printf("pipes made %d %d %d %d\n", fd1[0], fd1[1], fd2[0], fd2[1]);
-#endif
```
Comment (gapisback) [inline application_service/app_service.cc:502]: Moved this below, and enhanced output to report fd names clearly.

File: application_service/script [hunk 6]
```diff
@@ -17,54 +17,65 @@ mkdir $APP_SERVICE_DIR/provisioning
 
 # create key data
 cd $APP_SERVICE_DIR/provisioning
+
```
Comment (gapisback) [inline application_service/script:20]: Added blank line all over for improved readability.

File: sample_apps/run_example.sh [hunk 7]
```diff
@@ -516,6 +621,11 @@ function rm_non_git_files() {
 # ###########################################################################
 function build_utilities() {
     run_cmd
+    if [ "${SampleAppName}" = "simple_app_under_app_service" ]; then
+        echo "${Me}: Skipping ..."
```
Comment (gapisback) [inline sample_apps/run_example.sh:625]: Some steps are skipped while setting up the app, as they are supposed to have been "just-done" while `setup` is run for `application_service`.

**Review by jlmucb — APPROVED** [review]

**Review by yelvmw — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**